tools/buildgen/plugins/check_attrs.py

```diff
@@ -0,0 +1,126 @@
+        'boringssl': one_of((True, )),
+        'build_system': anything(),
+        'build': anything(),
+        'defaults': anything(),
+        'deps_linkage': one_of(('static',)),
+        'deps': anything(),
+        'dll': one_of((True, 'only')),
+        'filegroups': anything(),
+        'generate_plugin_registry': anything(),
+        'headers': anything(),
+        'language': one_of(('c', 'c++', 'csharp')),
+        'LDFLAGS': anything(),
+        'platforms': subset_of(('linux', 'mac', 'posix', 'windows')),
+        'public_headers': anything(),
+        'secure': one_of(('check', True, False)),
+        'src': anything(),
+        'vs_proj_dir': anything(),
+        'zlib': one_of((True, )),
+    },
+    'target': {
+        'args': anything(),
+        'benchmark': anything(),
+        'boringssl': one_of((True, )),
+        'build': anything(),
+        'ci_platforms': anything(),
+        'corpus_dirs': anything(),
+        'cpu_cost': anything(),
+        'defaults': anything(),
+        'deps': anything(),
+        'dict': anything(),
+        'exclude_configs': anything(),
+        'exclude_iomgrs': anything(),
+        'excluded_poll_engines': anything(),
+        'filegroups': anything(),
+        'flaky': one_of((True, False)),
+        'gtest': one_of((True, False)),
+        'headers': anything(),
+        'language': one_of(('c', 'c89', 'c++', 'csharp')),
+        'maxlen': anything(),
+        'platforms': subset_of(('linux', 'mac', 'posix', 'windows')),
+        'run': one_of((True, False)),
+        'secure': one_of(('check', True, False)),
+        'src': anything(),
+        'timeout_seconds': anything(),
+        'uses_polling': anything(),
+        'vs_proj_dir': anything(),
+        'zlib': one_of((True, )),
+    },
+}
+
+
+def check_attributes(entity, kind, errors):
+  attributes = VALID_ATTRIBUTE_KEYS_MAP[kind]
+  name = entity.get('name', anything())
+  for key, value in entity.items():
+    if key == 'name':
+      continue
+    validator = attributes.get(key)
+    if validator:
+      error = validator(value)
+      if error:
+        errors.append("{0}({1}) has an invalid value for '{2}': {3}".format(
+            name, kind, key, error))
+    else:
+      errors.append("{0}({1}) has an invalid attribute '{2}'".format(
+          name, kind, key))
+
+
+def mako_plugin(dictionary):
+  """The exported plugin code for check_attr.
+
+  This validates that filegroups, libs, and target can have only valid
+  attributes. This is mainly for preventing build.yaml from having
+  unnecessary and misleading attributes accidently.
+  """
+
+  errors = []
+  for filegroup in dictionary.get('filegroups', {}):
+    check_attributes(filegroup, 'filegroup', errors)
+  for lib in dictionary.get('libs', {}):
+    check_attributes(lib, 'lib', errors)
+  for target in dictionary.get('targets', {}):
+    check_attributes(target, 'target', errors)
+  #if errors:
+  #  raise Exception('\n'.join(errors))
```

Comment:
Why is this still commented out?

Reply:
oh my bad. it should be uncommented. thanks!